Improved support for working with paths that contain spaces, or legal but unusual characters like caret and ampersand.

bspratt opened the pull request:
Among other things, while running tests, when Skyline generates temporary filenames potentially troublesome characters like space, caret and ampersand are included in the filename.

I also changed the name of my test data downloads directory to include these characters.

We now escape such characters when passing arguments to msconvert's diaUmpire filter, as it requires nested quotes.

nickshulman commented:
I think we should write a class called "ArgumentList" which is a list of unescaped arguments and whose ToString() method produces a string will all of the escaped arguments in a row.

chambm commented on pwiz/analysis/spectrum_processing/SpectrumListFactory.cpp:
```diff
@@ -668,6 +688,26 @@ UsageInfo usage_precursorRecalculation = {"","This filter recalculates the precu
+        path = path.substr(1, path.length() - 2);
+    }
+    // Look for the ^ escape character
+    for (size_t escapePos = 0;;)
```
Is this necessary? I would expect the Windows shell to take the `^`s out (but not the escaped quotes). That's the case with that ugly titleMaker filter escaping I sent to you.

bspratt replied:
This is for dealing with what's between the nested quotes, which windows shell doesn't mess with. But I suppose that implies that we maybe didn't need to escape any of that stuff while constructing the command line. I'll have to review.

chambm replied:
You DO need to escape it when making the command-line in Skyline, but I'm hoping you don't need to process it inside the called exe (msconvert). Here's the titleMaker escaping even though it's inside the nested quotes:
```
--filter "titleMaker <RunId>.<ScanNumber>.<ScanNumber>.<ChargeState> File:"""^<SourcePath^>""", NativeID:"""^<Id^>""""
```
What msconvert sees: 
```
titleMaker <RunId>.<ScanNumber>.<ScanNumber>.<ChargeState> File:"<SourcePath>", NativeID:"<Id>"
```
But I concede it could be different if it's nested inside another level of quoting.

brendanx67 replied:
The escape characters should not show up inside the program. They are to direct the shell to give the correct characters to the program. If you handle them again side the program, you will be double-escaping, I think.

brendanx67 replied:
double-unescaping, I meant

chambm commented on pwiz_tools/Skyline/Util/UtilIO.cs:
```diff
@@ -946,6 +946,21 @@ public static string GetElapsedTimeExplanation(DateTime startTime, DateTime endT
+            if (Program.UnitTest)
+            {
+                // Introduce some potentially troublesome characters like space, caret, ampersand to test our handling
+                result = @"test c^ha&rs_" + result;
```
I like this function. How about some Unicode as well: 试验

bspratt replied:
Great idea!

brendanx67 commented:
> I think we should write a class called "ArgumentList" which is a list of unescaped arguments and whose ToString() method produces a string will all of the escaped arguments in a row.

Sounds okay to me, as long as we don't over-quote the arguments, e.g. putting quotes around every argument, which would work, but would look really lame when we show the command-lines to users.

bspratt commented:
> I think we should write a class called "ArgumentList" which is a list of unescaped arguments and whose ToString() method produces a string will all of the escaped arguments in a row.

Not following you, sorry

bspratt commented:
Oh, OK I get it. Though I'm not sure how to generalize the weirdness of the arguments to msconvert, with the nested quotes and all.

brendanx67 commented:
I think you would just have the functions "QuoteIfNecessary()" and
"EscapeIfNecessary()". If the Argument text is:

dia-umpire params="C:\Users\Test & Dev\Tmp\foobar.txt"

Then these functions would turn it into:

"dia-umpire params=\"C:\Users\Test ^& Dev\foobar.txt\""

Though, even the first part might be best handled by a
MsConvertFilterArgumentsList class with a FilterName and argument type
classes like NameAndValue.

Nick makes a good point that classes and polymorphism could do a lot here,
as they have for our own CommandArgs inside the Skyline Command Line
Interface.

On Thu, Feb 16, 2023 at 10:29 AM Brian Pratt ***@***.***>
wrote:

> Oh, OK I get it. Though I'm not sure how to generalize the weirdness of
> the arguments to msconvert, with the nested quotes and all.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ProteoWizard/pwiz/pull/2505#issuecomment-1433529943>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACYBWUGL4FWQY5OTNUJA6ETWXZWXDANCNFSM6AAAAAAU5RVL3Y>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

bspratt commented:
Just wanted to emphasize "I tried also adding some unicode (e.g. 试验, means "test") but it just breaks too many tests that involve unicode-ignorant 3rd party tools, mz5 files, etc."

I suppose we could add a flag for tests that we know won't work with unicode for reasons that are out of our control...

chambm commented:
> Just wanted to emphasize "I tried also adding some unicode (e.g. 试验, means "test") but it just breaks too many tests that involve unicode-ignorant 3rd party tools, mz5 files, etc."
> 
> I suppose we could add a flag for tests that we know won't work with unicode for reasons that are out of our control...

How many tests were broken that could be reasonable fixed (we control all the code) vs. those that can't? If the latter is a much greater number, then yeah it's probably not worth it.

bspratt commented:
At least a half dozen - MSFragger and anything using mz5 were the main culprits I observed. I can give it a shot.

brendanx67 commented on pwiz_tools/Shared/Common/SystemUtil/ProcessRunner.cs:
```diff
@@ -203,5 +204,15 @@ public IEnumerable<string> MessageLog()
         {
             return _messageLog;
         }
+
```
Feels like this should go into a new class named "ProcessEx". It is not really appropriate here. This class is about running another process not for asking questions about the currently running process.

bspratt replied:
OK, will do.

brendanx67 commented on pwiz_tools/Skyline/Util/Xml.cs:
```diff
@@ -961,6 +961,18 @@ public static string UnescapeNonPrintableChars(this string str)
             return sb.ToString();
         }
 
+        // Inspect a file path for characters that must be escaped for use in XML (currently just "&")
```
I could imagine putting this on PathEx also as EscapePathForXml() with EscapePathForCommandLine()... Or moving EscapePathForCommandLine() to ProcessRunner.EscapePath(). Feels unbalanced the way it is.

bspratt replied:
I can do that, sure.

brendanx67 reviewed (changes requested):
A few comments about where things live and definitely need to consider unescaping of ^ inside msconvert.exe. I think that escaping is all about getting the right text into the process. When that is right, then it should not contain any extra ^ characters that need to be removed.

bspratt commented on pwiz_tools/Skyline/Util/UtilIO.cs:
```diff
@@ -946,6 +946,23 @@ public static string GetElapsedTimeExplanation(DateTime startTime, DateTime endT
+        /// Like Path.GetRandFileName(), but in test context adds some legal but unusual characters for robustness testing
+        /// </summary>
+        /// <returns>a random filename</returns>
+        public static string GetRandomFileName()
```
OK

bspratt replied:
Hmm, problem is we don't normally get fancy with random filenames unless we're in a test. And that's not information PathEx has, over in pwiz.common

brendanx67 replied:
That makes me a little queasy that we are choosing code location to access internal test settings. I guess I would prefer something like a static on PathEx that is the RandomFile root and if it is not set, we just pass through to the normal Path.GetRandomFileName(). And the test framework sets it. That way these fairly low level classes don't need to reach outside themselves to look at test settings. Instead, the test framework pokes them into doing what it needs.

bspratt replied:
OK, I'll see what I can do.

bspratt commented:
I've set this as the Integration branch, and set BSPRATT-UW3 to run it as its Nightly.

bspratt commented:
@nickshulman @brendanx67 @chambm last chance to comment before I merge this, with this commit message:

**Improved support for working with paths that contain spaces, or legal but unusual characters like caret and ampersand, and for the issue of stray files being left behind by various tests and 3rd party tools**

* Made changes to the Skyline automated testing scheme to improve our support for users with windows usernames containing spaces an other legal but unusual characters. This is achieved by setting the TMP environment variable to point at a directory whose name contains such characters. Normally that value is based on the user name, so even on a system with username "ihavenospacesinmyname" we can test running with a TMP dir with a name that contains space, caret, and ampersand.

* We now properly escape quotes when passing arguments to msconvert's diaUmpire filter, as it requires nested quotes.

* Additional cleanup for Crux pepXML output that doesn't properly escape the & in filenames

* Added a custom attribute for tests that don't do well with unicode paths, usually due to mz5 or 3rd party tool usage. Tests without the NoUnicodeTesting(string reason) will run with TMP and results directories whose names include unicode characters.

* Check for cleanup of our temp directory on test exit

* Explicitly clean up MSAmanda's tempfiles, which are left on purpose but which violate the general expectations of test cleanup.  Added a couple of TODO comments about how we need to think more about MSAmanda leaves tempfiles around.

* ProcessRunner.Run now accepts an optional argument "forceTempfilesCleanup" which, when true, runs the process with an altered environment in which TMP is a subdirectory of the original temp directory. Afterward the directory is deleted.

* added pwiz.Common.SystemUtil.TempDir class to make it easier to clean up after things like ExcelReaderFactory that leave tempfiles behind

* Custom tools that write to the immediate window weren't cleaning up their temp files.

* Do a better job of cleaning up tempfiles when an MSFragger search is interrupted

brendanx67 commented on pwiz_tools/Skyline/TestPerf/DdaTutorialTest.cs:
```diff
@@ -41,7 +41,7 @@ namespace TestPerf
     public class DdaTutorialTest : AbstractFunctionalTest
     {
-        [TestMethod, NoParallelTesting(TestExclusionReason.RESOURCE_INTENSIVE)]
+        [TestMethod, NoParallelTesting(TestExclusionReason.RESOURCE_INTENSIVE), NoUnicodeTesting("MsFragger does not handle unicode paths")]
```
These should probably be constants on TestExclusionReason. They appear to get repeated.

brendanx67 commented on pwiz_tools/Skyline/TestRunnerLib/RunTests.cs:
```diff
@@ -332,6 +338,34 @@ public bool Run(TestInfo test, int pass, int testNumber, string dmpDir, bool hea
                     Environment.SetEnvironmentVariable(@"SKYLINE_TESTER_PARALLEL_CLIENT_ID", ParallelClientId); // Accessed in pwiz_tools\Skyline\Util\Util.cs
                 }
 
+                // Set the temp file path to something peculiar - helps guarantee support for
```
Put this in a function instead of making a BIG function bigger.

chambm replied:
If this gets moved to a function it needs to be a functor with a Dispose call that resets the TMP variable (as is done later in this function). But having that environment-restoring behavior is probably a good idea anyway if the code throws between the time TMP is redirected and when it is restored.

brendanx67 replied:
Or the function needs to return tmpTestDir?

brendanx67 commented on pwiz_tools/Skyline/TestRunnerLib/RunTests.cs:
```diff
@@ -517,29 +554,55 @@ public bool Run(TestInfo test, int pass, int testNumber, string dmpDir, bool hea
-        private List<string> CleanUpTestDir()
+        private List<string> CleanUpTestDir(string tmpTestDir, bool final)
         {
+            CleanupMSAmandaTmpFiles();
```
Put this in the tests that need it or put a TODO comment on it, if these still get created all the time.

bspratt replied:
There's already one in the function itself, FWIW, I'll add another here.

brendanx67 commented on pwiz_tools/Skyline/TestRunnerLib/RunTests.cs:
```diff
@@ -517,29 +554,55 @@ public bool Run(TestInfo test, int pass, int testNumber, string dmpDir, bool hea
             {
-                allEntries.AddRange(Directory.EnumerateFileSystemEntries(TestContext.TestDir));
-                if (allEntries.Count > 0)
+                foreach (var dir in new []{TestContext.TestDir, tmpTestDir})
```
Use sub-functions instead of loops around an existing body of code.

bspratt replied:
Fair enough. Updated.

brendanx67 reviewed (changes requested):
Definitely getting there! Lots of issues solved. Still have a few suggestions.

brendanx67 commented on pwiz_tools/Shared/Common/SystemUtil/ProcessRunner.cs:
```diff
@@ -53,23 +53,48 @@ public class ProcessRunner : IProcessRunner
 
             _messageLog.Clear();
 
+            if (forceTempfilesCleanup)
```
This appears to be another place where a sub-function might improve readability rather than adding a new chunk of code to an already large function dedicated to a single issue.

bspratt replied:
No problem.

brendanx67 commented on pwiz_tools/Skyline/TestRunnerLib/RunTests.cs:
```diff
@@ -517,34 +530,98 @@ public bool Run(TestInfo test, int pass, int testNumber, string dmpDir, bool hea
-                allEntries.AddRange(Directory.EnumerateFileSystemEntries(TestContext.TestDir));
-                if (allEntries.Count > 0)
+                CleanupAbandonedFiles(TestContext.TestDir, true, abandonedFilesList);
+                CleanupAbandonedFiles(tmpTestDir, !final, abandonedFilesList);
```
I guess I wonder about this placement. It would seem to imply that we are fine with leaving temp files around when _cleanupLevelAll is false, as we really are with leaving around the files tests produce, since they can be very helpful artifacts when a test fails. I could imagine we would want to see the temp files go away no matter what the clean-up level. Are we expecting normal use to leave temp files around?

chambm replied:
It would not be atypical program behavior at least. Just look at all the cruft that builds up in %TEMP%. :) But I think if Skyline is creating a temp file that is supposed to be cleaned up during execution but maybe isn't due to a bug or error, it's ok to clean it up always. If we want to keep a file around for understanding why an error is happening, it's better to have it be a non-temp file (like I will move the DDA params files to the output directory instead of TEMP).

brendanx67 replied:
Right. I think that means that the tmpTestDir check should go outside the if block, so that we check for clean-up regardless of whether the test settings are set to clean up all files.

bspratt replied:
I agree.

brendanx67 reviewed (commented):
Thanks for the improvements. Couple last issues.

bspratt commented:
OK, most recent push should address these concerns.
